run-ws.py

```diff
@@ -0,0 +1,128 @@
+import os
+import json
+import socket
+import signal
+import asyncio
+import logging
+import argparse
+import websockets
+
+from jsonschema import validate
+from chainlink import Chainlink
+
+import grader.api_keys as api_keys
+from grader.definitions import GRADING_JOB_DEF
+
+from config import *
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger()
+
+
+async def exec_job(job):
+    job_id = job[api_keys.GRADING_JOB_ID]
+    stages = job[api_keys.STAGES]
+
+    logger.info("starting job {}".format(job_id))
+
+    # execute job
+    try:
+        chain = Chainlink(stages, workdir=os.getcwd())
+        job_results = await chain.run_async({})
+    except Exception as ex:
+        logger.critical("grading job failed with exception:\n{}", ex)
+        job_results = [
+            {
+                "logs": {
+                    "stdout": b"the container crashed",
+                    "stderr": bytes(str(ex), "utf-8"),
+                },
+                "success": False,
+            }
+        ]
+
+    job_stdout = "\n".join(
+        [r["logs"]["stdout"].decode("utf-8") for r in job_results]
+    )
+    job_stderr = "\n".join(
+        [r["logs"]["stderr"].decode("utf-8") for r in job_results]
+    )
+
+    for r in job_results:
+        del r["logs"]
+
+    logger.info("finished job {}".format(job_id))
+
+    logger.info("job stdout:\n" + job_stdout)
+    logger.info("job stderr:\n" + job_stderr)
+
+    return {
+        api_keys.RESULTS: job_results,
+        api_keys.SUCCESS: job_results[-1]["success"],
+        api_keys.LOGS: {"stdout": job_stdout, "stderr": job_stderr},
+        api_keys.GRADING_JOB_ID: job_id,
+    }
+
+async def run(token, worker_id):
+    url = "{}://{}:{}{}{}/{}".format(
+        "wss" if USE_SSL else "ws",
+        API_HOSTNAME, API_PORT,
+        API_PROXY, WORKER_WS_ENDPOINT,
+        worker_id,
+    )
+
+    headers = {api_keys.AUTH: "Bearer {}".format(token)}
+    hostname = socket.gethostname()
+
+    async with websockets.connect(url, extra_headers=headers) as ws:
+        # poll job
+        try:
+            await ws.send(json.dumps({
+                "type": "register",
+                "args": {"hostname": hostname},
+            }))
+
+            ack = json.loads(await ws.recv())
+
+            if not ack["success"]:
+                raise Exception("failed to register")
+
+            while True:
+                job = json.loads(await ws.recv())
+
+                validate(instance=job, schema=GRADING_JOB_DEF)
```

Comment:
If you look at the source code of `validate`, it can raise `ValidationError` if the instance is invalid or `SchemaError` if the schema itself is invalid. I think we should be catching these below and logging appropriate messages.